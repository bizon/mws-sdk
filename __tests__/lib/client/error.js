const {HTTPError} = require('got')

const MWSError = require('../../../lib/client/error')

const {parseStr} = require('../../../lib/client/parsers/base')

describe('lib.client.error', () => {
  it('should rethrow if httpError is not a valid got HTTPError', () => {
    const errors = [
      new Error(),
      new TypeError()
    ]

    for (const error of errors) {
      expect(
        () => new MWSError(error)
      ).toThrow(error)
    }
  })

  it('should fail if resource or method is not defined', () => {
    const tests = [
      () => new MWSError(new HTTPError({})),
      () => new MWSError(new HTTPError({}), 'Order')
    ]

    for (const test of tests) {
      expect(test).toThrow('The resource and method parameters are required')
    }
  })

  it('should create a MWSError and parse the response body', () => {
    const error = new MWSError(
      new HTTPError({
        statusCode: 400,
        statusMessage: 'Something wrong happened',
        body: `<ErrorResponse>
          <Error>
            <Type>Foo</Type>
            <Code>Bar</Code>
            <Message>This is a message</Message>
          </Error>
          <ResponseMetadata>
            <RequestId>some_request_id</RequestId>
          </ResponseMetadata>
        </ErrorResponse>`
      }),
      'ResourceName',
      'Function'
    )

    expect(error.message).toBe('ResourceName.Function error: Response code 400 (Something wrong happened)')
    expect(error.body).toMatchSnapshot()
  })

  it('should create a MWSError and parse the response body using a custom error response parser', () => {
    const error = new MWSError(
      new HTTPError({
        statusCode: 400,
        statusMessage: 'Something wrong happened',
        body: `<ErrorResponse xmlns="https://mws.amazonservices.com/Sellers/2011-07-01">
          <Foo>Hello</Foo>
          <Bar>World!</Bar>
        </ErrorResponse>`
      }),
      'Sellers',
      'GetSomething',
      (key, node) => ({
        foo: parseStr(`${key}/sellers:Foo`, node),
        bar: parseStr(`${key}/sellers:Bar`, node)
      })
    )

    expect(error.message).toBe('Sellers.GetSomething error: Response code 400 (Something wrong happened)')
    expect(error.body).toMatchSnapshot()
  })
})
