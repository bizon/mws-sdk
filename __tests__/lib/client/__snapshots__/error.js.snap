// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib.client.error should create a MWSError and parse the response body 1`] = `
Object {
  "error": Object {
    "code": "Bar",
    "detail": null,
    "message": "This is a message",
    "type": "Foo",
  },
  "responseMetadata": Object {
    "requestId": "some_request_id",
  },
}
`;

exports[`lib.client.error should create a MWSError and parse the response body using a custom error response parser 1`] = `
Object {
  "bar": "World!",
  "foo": "Hello",
}
`;

exports[`lib.client.error should not allow updating error properties 1`] = `
Object {
  "action": "Function",
  "body": Object {
    "error": Object {
      "code": null,
      "detail": null,
      "message": "This is a message",
      "type": null,
    },
    "responseMetadata": Object {
      "requestId": null,
    },
  },
  "marketplaces": Array [
    Object {
      "advertisingApiDomain": "advertising-api-eu.amazon.com",
      "code": "de",
      "domain": "amazon.de",
      "id": "A1PA6795UKMFR9",
      "imagesDomain": "images-eu.ssl-images-amazon.com",
      "mwsDomain": "mws-eu.amazonservices.com",
      "name": "Germany",
      "region": "eu",
      "vendorId": "A3JWKAKR8XB7XF",
    },
    Object {
      "advertisingApiDomain": "advertising-api-eu.amazon.com",
      "code": "es",
      "domain": "amazon.es",
      "id": "A1RKKUPIHCS9HS",
      "imagesDomain": "images-eu.ssl-images-amazon.com",
      "mwsDomain": "mws-eu.amazonservices.com",
      "name": "Spain",
      "region": "eu",
      "vendorId": "A1AT7YVPFBWXBL",
    },
    Object {
      "advertisingApiDomain": "advertising-api-eu.amazon.com",
      "code": "fr",
      "domain": "amazon.fr",
      "id": "A13V1IB3VIYZZH",
      "imagesDomain": "images-eu.ssl-images-amazon.com",
      "mwsDomain": "mws-eu.amazonservices.com",
      "name": "France",
      "region": "eu",
      "vendorId": "A1X6FK5RDHNB96",
    },
    Object {
      "advertisingApiDomain": "advertising-api-eu.amazon.com",
      "code": "it",
      "domain": "amazon.it",
      "id": "APJ6JRA9NG5V4",
      "imagesDomain": "images-eu.ssl-images-amazon.com",
      "mwsDomain": "mws-eu.amazonservices.com",
      "name": "Italy",
      "region": "eu",
      "vendorId": "A11IL2PNWYJU7H",
    },
    Object {
      "advertisingApiDomain": "advertising-api-eu.amazon.com",
      "code": "uk",
      "domain": "amazon.co.uk",
      "id": "A1F83G8C2ARO7P",
      "imagesDomain": "images-eu.ssl-images-amazon.com",
      "mwsDomain": "mws-eu.amazonservices.com",
      "name": "United Kingdom",
      "region": "eu",
      "vendorId": "A3P5ROKL5A1OLE",
    },
    Object {
      "code": "tr",
      "domain": "amazon.com.tr",
      "id": "A33AVAJ2PDY3EV",
      "imagesDomain": "images-eu.ssl-images-amazon.com",
      "mwsDomain": "mws-eu.amazonservices.com",
      "name": "Turkey",
      "region": "eu",
      "vendorId": "A1UNQM1SR2CHM",
    },
    Object {
      "code": "eg",
      "id": "ARBP9OOSHTCHU",
      "imagesDomain": "images-eu.ssl-images-amazon.com",
      "mwsDomain": "mws-eu.amazonservices.com",
      "name": "Egypt",
      "region": "eu",
    },
    Object {
      "code": "sa",
      "domain": "amazon.sa",
      "id": "A17E79C6D8DWNP",
      "imagesDomain": "images-eu.ssl-images-amazon.com",
      "mwsDomain": "mws-eu.amazonservices.com",
      "name": "Saudi Arabia",
      "region": "eu",
      "vendorId": "A2XPWB6MYN7ZDK",
    },
    Object {
      "code": "nl",
      "domain": "amazon.nl",
      "id": "A1805IZSGTT6HS",
      "imagesDomain": "images-eu.ssl-images-amazon.com",
      "mwsDomain": "mws-eu.amazonservices.com",
      "name": "Netherlands",
      "region": "eu",
      "vendorId": "A17D2BRD4YMT0X",
    },
    Object {
      "code": "se",
      "domain": "amazon.se",
      "id": "A2NODRKZP88ZB9",
      "imagesDomain": "images-eu.ssl-images-amazon.com",
      "mwsDomain": "mws-eu.amazonservices.com",
      "name": "Sweden",
      "region": "eu",
    },
  ],
  "mwsDomain": "mws-eu.amazonservices.com",
  "name": "MWSError",
  "resource": "ResourceName",
  "sellerId": "sellerId",
}
`;
